### PR TITLE
Add model examples to docs

### LINKED ISSUE #590: Add examples for all models in documentation
Create example code for every model we support using low-level blocks.

- [x] BYOL
- [x] Barlow Twins
- [x] MoCo
- [x] NNCLR
- [x] SimCLR
- [x] SimSiam
- [x] SwaV

Relates to #493

### PR DESCRIPTION
**Pull request by guarin**
Closes #590 

Adds low-level model examples for all models we currently support.

Example pdf:
[moco_example.pdf](https://github.com/lightly-ai/lightly/files/7617494/moco_example.pdf)

### CODE REVIEW

File: examples/pytorch/barlowtwins.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import torch
+from torch import nn
+import torchvision
+
+import lightly
+from lightly.models.modules import BarlowTwinsProjectionHead
```
Comment (philippmwirth) [inline examples/pytorch/barlowtwins.py:6]: It's a little strange that you import this directly but use `lightly.loss.BarlowTwins` and `lightly.data.LightlyDataset`

File: examples/pytorch/barlowtwins.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    num_workers=8,
+)
+
+criterion = lightly.loss.BarlowTwinsLoss()
+optimizer = torch.optim.SGD(model.parameters(), lr=0.06)
+
+print("Starting Training")
+for epoch in range(10):
+    total_loss = 0
+    for (x0, x1), file_name, label in dataloader:
```
Comment (philippmwirth) [inline examples/pytorch/barlowtwins.py:50]: if we don't use the label maybe let's just ignore it? Might be more confusing than helpful

File: examples/pytorch/barlowtwins.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        x0 = x0.to(device)
+        x1 = x1.to(device)
+        z0 = model(x0)
+        z1 = model(x1)
+        loss = criterion(z0, z1)
+        total_loss += loss
```
Comment (philippmwirth) [inline examples/pytorch/barlowtwins.py:56]: detach?

File: examples/pytorch/byol.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
+import torch
+from torch import nn
+import torchvision
+import copy
+
+import lightly
+from lightly.models.modules import BYOLProjectionHead
```
Comment (philippmwirth) [inline examples/pytorch/byol.py:7]: I'll just write "consistent imports" everywhere now since it seems there's the same problem everywhere

File: examples/pytorch/byol.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    num_workers=8,
+)
+
+criterion = lightly.loss.NegativeCosineSimilarity()
+optimizer = torch.optim.SGD(model.parameters(), lr=0.06)
+
+print("Starting Training")
+for epoch in range(10):
+    total_loss = 0
+    for (x0, x1), file_name, label in dataloader:
```
Comment (philippmwirth) [inline examples/pytorch/byol.py:64]: remove filename and label?

File: examples/pytorch/byol.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        x0 = x0.to(device)
+        x1 = x1.to(device)
+        p0, z1 = model(x0, x1)
+        p1, z0 = model(x1, x0)
```
Comment (philippmwirth) [inline examples/pytorch/byol.py:68]: wouldn't it be nicer to separate the forward passes here? (I know this is how we did it so far but there's a reason we want to change this 😉 ), e.g.

```python
for epoch in range(10):
    total_loss = 0
    for (x0, x1), _, _ in dataloader:
        x0 = x0.to(device)
        x1 = x1.to(device)
        p0 = model(x0)
        p1 = model(x1)
        z0 = model.forward_momentum(x0)
        z1 = model.forward_momentum(x1) # the name is up to debate
        # continue as before
```

File: examples/pytorch/moco.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
+import torch
+from torch import nn
+import torchvision
+import copy
+
+import lightly
+from lightly.models.modules import MoCoProjectionHead
```
Comment (philippmwirth) [inline examples/pytorch/moco.py:7]: consistent imports

File: examples/pytorch/moco.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        self.backbone = backbone
+        self.projection_head = MoCoProjectionHead(512, 512, 128)
+
+        self.backbone_momentum = copy.deepcopy(self.backbone)
+        self.projection_head_momentum = copy.deepcopy(self.projection_head)
+
+        deactivate_requires_grad(self.backbone_momentum)
+        deactivate_requires_grad(self.projection_head_momentum)
+
+    def forward(self, x_q, x_k):
```
Comment (philippmwirth) [inline examples/pytorch/moco.py:25]: same argument here as for BYOL: wouldn't it be better to split the forward passes for the encoder and the momentum encoder?

File: examples/pytorch/nnclr.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
+import torch
+from torch import nn
+import torchvision
+
+import lightly
+from lightly.models.modules import NNCLRProjectionHead
+from lightly.models.modules import NNCLRPredictionHead
```
Comment (philippmwirth) [inline examples/pytorch/nnclr.py:7]: consistent imports

File: examples/pytorch/simclr.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
+import torch
+from torch import nn
+import torchvision
+
+import lightly
+from lightly.models.modules import SimCLRProjectionHead
```
Comment (philippmwirth) [inline examples/pytorch/simclr.py:6]: consistent imports

File: examples/pytorch/simsiam.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
+import torch
+from torch import nn
+import torchvision
+
+import lightly
+from lightly.models.modules import SimSiamProjectionHead
+from lightly.models.modules import SimSiamPredictionHead
```
Comment (philippmwirth) [inline examples/pytorch/simsiam.py:7]: consistent imports

File: examples/pytorch/swav.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+from lightly.models.modules import SwaVProjectionHead
+from lightly.models.modules import SwaVPrototypes
+
+
+class SwaV(nn.Module):
+    def __init__(self, backbone):
+        super().__init__()
+        self.backbone = backbone
+        self.projection_head = SwaVProjectionHead(512, 512, 128)
+        self.prototypes = SwaVPrototypes(128, 512)
```
Comment (philippmwirth) [inline examples/pytorch/swav.py:15]: maybe a comment here what this means - it means we use 512 prototypes, right?

File: examples/pytorch/swav.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
+import torch
+from torch import nn
+import torchvision
+
+import lightly
+from lightly.models.modules import SwaVProjectionHead
```
Comment (philippmwirth) [inline examples/pytorch/swav.py:6]: consistent imports

File: examples/pytorch_lightning/barlowtwins.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
+import torch
+from torch import nn
+import torchvision
+import pytorch_lightning as pl
+
+import lightly
+from lightly.models.modules import BarlowTwinsProjectionHead
```
Comment (philippmwirth) [inline examples/pytorch_lightning/barlowtwins.py:7]: consistent imports

File: examples/pytorch_lightning/byol.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
+import torch
+from torch import nn
+import torchvision
+import copy
+import pytorch_lightning as pl
+
+import lightly
+from lightly.models.modules import BYOLProjectionHead
```
Comment (philippmwirth) [inline examples/pytorch_lightning/byol.py:8]: consistent imports

File: examples/pytorch_lightning/byol.py [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+        self.backbone_momentum = copy.deepcopy(self.backbone)
+        self.projection_head_momentum = copy.deepcopy(self.projection_head)
+
+        deactivate_requires_grad(self.backbone_momentum)
+        deactivate_requires_grad(self.projection_head_momentum)
+
+        self.criterion = lightly.loss.NegativeCosineSimilarity()
+
+    def forward(self, x0, x1):
+        y0 = self.backbone(x0).flatten(start_dim=1)
```
Comment (philippmwirth) [inline examples/pytorch_lightning/byol.py:30]: same argument as above, we could split the forward pass

File: examples/pytorch_lightning/moco.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
+import torch
+from torch import nn
+import torchvision
+import copy
+import pytorch_lightning as pl
+
+import lightly
+from lightly.models.modules import MoCoProjectionHead
+from lightly.models.utils import deactivate_requires_grad
```
Comment (philippmwirth) [inline examples/pytorch_lightning/moco.py:9]: consistent ipmorts

File: examples/pytorch_lightning/moco.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+        self.backbone_momentum = copy.deepcopy(self.backbone)
+        self.projection_head_momentum = copy.deepcopy(self.projection_head)
+
+        deactivate_requires_grad(self.backbone_momentum)
+        deactivate_requires_grad(self.projection_head_momentum)
+
+        self.criterion = lightly.loss.NTXentLoss(memory_bank_size=4096)
+
+    def forward(self, x_query, x_key):
+        query = self.backbone(x_query).flatten(start_dim=1)
```
Comment (philippmwirth) [inline examples/pytorch_lightning/moco.py:29]: split?

File: examples/pytorch_lightning/nnclr.py [hunk 19]
```diff
@@ -0,0 +1,65 @@
+import torch
+from torch import nn
+import torchvision
+import pytorch_lightning as pl
+
+import lightly
+from lightly.models.modules import NNCLRProjectionHead
+from lightly.models.modules import NNCLRPredictionHead
```
Comment (philippmwirth) [inline examples/pytorch_lightning/nnclr.py:8]: consistent imports

File: examples/pytorch_lightning/simclr.py [hunk 20]
```diff
@@ -0,0 +1,56 @@
+import torch
+from torch import nn
+import torchvision
+import pytorch_lightning as pl
+
+import lightly
+from lightly.models.modules import SimCLRProjectionHead
```
Comment (philippmwirth) [inline examples/pytorch_lightning/simclr.py:7]: consistent imports

File: examples/pytorch_lightning/simsiam.py [hunk 21]
```diff
@@ -0,0 +1,62 @@
+import torch
+from torch import nn
+import torchvision
+import pytorch_lightning as pl
+
+import lightly
+from lightly.models.modules import SimSiamProjectionHead
+from lightly.models.modules import SimSiamPredictionHead
```
Comment (philippmwirth) [inline examples/pytorch_lightning/simsiam.py:8]: consistent imports

File: examples/pytorch_lightning/swav.py [hunk 22]
```diff
@@ -0,0 +1,64 @@
+import torch
+from torch import nn
+import torchvision
+import pytorch_lightning as pl
+
+import lightly
+from lightly.models.modules import SwaVProjectionHead
```
Comment (philippmwirth) [inline examples/pytorch_lightning/swav.py:7]: consistent imports

**Review by philippmwirth — COMMENTED** [review]
First of all, great work!!! Looking forward to seeing the examples released :) 

I left some feedback.

File: examples/pytorch/barlowtwins.py (hunk 1 above)
Comment (guarin) [reply examples/pytorch/barlowtwins.py:6]: Updated :)

File: examples/pytorch/barlowtwins.py (hunk 2 above)
Comment (guarin) [reply examples/pytorch/barlowtwins.py:50]: Ok, removed them.

File: examples/pytorch/barlowtwins.py (hunk 3 above)
Comment (guarin) [reply examples/pytorch/barlowtwins.py:56]: Afaik there is no stop gradient in barlow twins.

File: examples/pytorch_lightning/moco.py (hunk 18 above)
Comment (guarin) [reply examples/pytorch_lightning/moco.py:29]: What do you mean by split?

File: examples/pytorch/barlowtwins.py (hunk 3 above)
Comment (guarin) [reply examples/pytorch/barlowtwins.py]: Done :)

File: examples/pytorch/barlowtwins.py [hunk 23]
```diff
@@ -55,7 +55,7 @@ def forward(self, x):
         z0 = model(x0)
         z1 = model(x1)
         loss = criterion(z0, z1)
-        total_loss += loss
+        total_loss += loss.detach().detach()
```
Comment (philippmwirth) [inline examples/pytorch/barlowtwins.py:58]: double detach 😄

Comment (guarin) [reply examples/pytorch/barlowtwins.py]: Fixed :)

**Review by philippmwirth — APPROVED** [review]
